### ISSUE TITLE
"bad array new length" error when using PROPACK

### ISSUE THREAD

**Issue by ma-lima** [dwelter/pestpp#8]
If I use the PROPACK SVD package, I get a "bad array new length" error when PEST++ starts computing the upgrade vectors (after it calculates the Jacobian).  If I use the default SVD package the error does not occur and PEST++ runs normally to completion.

Here are some selections from the output file:

```
PEST++ run mode:- 
   regularization (with a "z")

Case dimensions:- 
    Number of parameters = 1669
    Number of adjustable parameters = 1669
    Number of observations = 1114
    Number of prior estimates = 36525

PEST Control Information
    relparmax = 10
    facparmax = 10
    facorig = 0.001
    phiredswh = 0.1
    noptmax = 25
    phiredstp = 0.01
    nphistp = 3
    nphinored = 3
    relparstp = 0.01
    nrelpar = 3

PEST++ Options
    n_iter_base = 1                   
    n_iter_super = 3                   
    max_n_super = 2000                
    super eigthres = 5e-07               
    svd pack = 1                   
    auto norm = 4                   
    super relparmax = 0.1                 
    max super frz iter = 5                   
    mat inv = 1                   
    max run fail = 3                   
    max reg iter = 20                  
    lambdas = 
            0.1
              1
             10
            100
           1000
    uncertainty flag = 1                   
    parameter covariance file =                     
    prediction names = 

PEST SVD Information
    maxsing = 1669
    eigthresh = 5e-07

   -----    Starting PEST++ Iterations    ----    
OPTIMISATION ITERATION NUMBER: 1

  Iteration type: base parameter solution
    SVD Package: PROPACK
    Matrix Inversion: "Jt Q J"
    Model calls so far : 0

    Current regularization weight factor                      : 0.01
    Starting phi for this iteration                     Total : 89757.9
    Starting measurement phi for this iteration         Total : 89757.9
    Starting regularization phi for this iteration      Total : 0
    Contribution to phi from observation group    "regul_kp1" : 0
    Contribution to phi from observation group        "head1" : 89757.9
    Contribution to phi from observation group        "regul" : 0
    Number of terms in the jacobian equal to zero: 61143104 / 62819491 (97%)

bad array new length
FATAL ERROR: base parameter run failed.
```

If I don't use PROPACK, the output file continues as follows:

```
Number of terms in the jacobian equal to zero: 61143104 / 62819491 (97%)


    ---  Solving for regularization weight factor   ---   
    Starting regularization weight factor      : 0.01
    Starting measurement objective function    : 89757.9
    Starting regularization objective function : 0

    Target measurement objective function      : 8975.79
```

The only change between the two is the use of PROPACK.

**Comment by jtwhite79** [comment]
Thanks for the heads up.  I can reproduce this problem but only with larger-scale problems (npar > 1000ish), which makes debugging painful...working on it.

**Comment by jtwhite79** [comment]
I believe this issue has been solved in the most recent commits to the master branch.  Dave found the issue - it was related to the size of the working vector that propack was trying to allocate.  I tested the fix on a problem with 1000s of pars and 10,000s obs and seems to be working - as a result of limited the size of this working vector (and limited the number of internal propack SVD iterations) you might get some screen output warning about ``Maximum dimension of  Krylov subspace exceeded prior to convergence`` - this can likely be ignored. Let us know if it's not working for you.

Also, we just implemented a randomized SVD algorithm from [https://github.com/ntessore/redsvd-h](https://github.com/ntessore/redsvd-h).  It can be activate with ``++svd_pack(redsvd)``.  Some initial testing shows that it can be very fast, especially with judicious use of the ``max_sing`` control parameter.  For example, it can find the first 50 singular components of a 21,000 obs X 2,700 par (16.5 million nonzeros in the jacobian) problem in 5 sec (about the same as propack). As problem sizes grow, I would expect this SVD solver to be faster than propack.
